vmm/src/console_devices.rs

```diff
@@ -0,0 +1,300 @@
+            vmconfig.console.file = Some(path.clone());
+            vmm.console_resize_pipe = Some(
+                listen_for_sigwinch_on_tty(
+                    sub_fd,
+                    &vmm.seccomp_action,
+                    vmm.hypervisor.hypervisor_type(),
+                )
+                .map_err(ConsoleDeviceError::StartSigwinchListener)?,
+            );
+        }
+        ConsoleOutputMode::Tty => {
+            // Duplicating the file descriptors like this is needed as otherwise
+            // they will be closed on a reboot and the numbers reused
+
+            // SAFETY: FFI call to dup. Trivially safe.
+            let stdout = unsafe { libc::dup(libc::STDOUT_FILENO) };
+            if stdout == -1 {
+                return vmm_sys_util::errno::errno_result().map_err(ConsoleDeviceError::DupFd);
+            }
+            // SAFETY: stdout is valid and owned solely by us.
+            let stdout = unsafe { File::from_raw_fd(stdout) };
+
+            // SAFETY: FFI call. Trivially safe.
+            if unsafe { libc::isatty(libc::STDOUT_FILENO) } == 1 {
+                vmm.console_resize_pipe = Some(
+                    listen_for_sigwinch_on_tty(
+                        stdout.try_clone().unwrap(),
+                        &vmm.seccomp_action,
+                        vmm.hypervisor.hypervisor_type(),
+                    )
+                    .map_err(ConsoleDeviceError::StartSigwinchListener)?,
+                );
+            }
+
+            // Make sure stdout is in raw mode, if it's a terminal.
+            set_raw_mode(&stdout, vmm.original_termios_opt.clone())?;
+            console_info.console_main_fd = Some(stdout.into_raw_fd());
+        }
+        ConsoleOutputMode::Null | ConsoleOutputMode::Socket | ConsoleOutputMode::Off => {}
+    }
+
+    match vmconfig.serial.mode {
+        ConsoleOutputMode::File => {
+            let file = File::create(vmconfig.serial.file.as_ref().unwrap())
+                .map_err(ConsoleDeviceError::CreateConsoleDevice)?;
+            console_info.serial_main_fd = Some(file.into_raw_fd());
+        }
+        ConsoleOutputMode::Pty => {
+            let (main_fd, sub_fd, path) =
+                create_pty().map_err(ConsoleDeviceError::CreateConsoleDevice)?;
+            console_info.serial_main_fd = Some(main_fd.into_raw_fd());
+            set_raw_mode(&sub_fd.as_raw_fd(), vmm.original_termios_opt.clone())?;
+            vmconfig.serial.file = Some(path.clone());
+        }
+        ConsoleOutputMode::Tty => {
+            // During vm_shutdown, when serial device is closed, FD#2(STDOUT)
+            // will be closed and FD#2 could be reused in a future boot of the
+            // guest by a different file.
+            // To ensure FD#2 always points to STANDARD OUT, a `dup` of STDOUT
+            // is passed to serial device. Doing so, even if the serial device
+            // were to be closed, FD#2 will continue to point to STANDARD OUT.
```

Comment:
I don't get the reasoning in the first paragraph. I get the second paragraph.

If fd 2 will be reused, how did it work before your change?

BTW please put a blank line between the paragraphs.

Reply:
The prior version has logic to saved the ptys from previous boot as a `File` using `PtyPair`. I believe this prevents the FD#2 from being closed and reused. 

In this new version, to keep the code flow simple, the FDs are re-opened using `pre_create_console_devices` just after `vm.shutdown` in `vm_reboot`. This new flow allows FD#2 to be closed and reused if it is not `dup`ed.